text/0000-cargo-templates.md

```diff
@@ -0,0 +1,275 @@
+- Feature Name: `cargo_new_project_from_template`
+- Start Date: 2020-05-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/cargo#3506](https://github.com/rust-lang/cargo/issues/3506)
+
+# Summary
+
+Add the ability to reference a template when running `cargo new` or `cargo init`
+
+# Motivation
+
+> We're doing this to make it "easy for frameworks to ship starter templates to users to give users the dependencies, stubs etc that they need" - @withoutboats 
+
+[Source](https://internals.rust-lang.org/t/pre-rfc-cargo-new-templates/12060/10?u=jsjoeio.)
+
+The expected outcome is to:
+
+* support templates that live on the crates.io registry inside a directory called `templates` at the root of a crate
+
+# Guide-level explanation
+
+Let's create a new project using Cargo and a template from the registry. For example, we'll try this using the `hello-world` template in the `rocket` crate. To use this run the following:
+
+```
+cargo new my-rocket-app --template rocket/hello-world
+```
+
+Note: this option is also supported with the `init` command:
+
+```
+cargo init my-rocket-app --template rocket/hello-world
+```
+
+This will create a new cargo project called `my-rocket-app` using the `hello-world` template.
+
+To view other `rocket` templates, see the `templates` directory in the root of the `rocket` repository on GitHub.
+
+# Reference-level explanation
+
+`cargo new` and `cargo init` will work the same way as usual. The added feature will be the ability to use a template that exists within a package on the [crates.io](http://crates.io) registry.
+
+Templates can be created by any package as long as they follow these conventions:
+
+* the package is on the [crates.io](http://crates.io) registry
+* the template lives inside the `templates` directory, which lives at the root of the crate
+
+Example:
+
+```
+my-crate
+- templates
+    - hello-world
+    - basic-app
+    - advanced-app 
+```
+
+If this were a real crate, the following would be valid ways to initialize a project with the `my-crate` templates:
+
+* `cargo new my-new-app --template my-crate/hello-world `
+* `cargo new my-new-app --template my-crate/basic-app`
+* `cargo new my-new-app --template  my-crate/advanced-app`
+
+Note: the same examples could be used with `cargo init` as well.
+
+To summarize, the following is the syntax for using a template
+
+* `cargo new <project-name> --template <crate-name>/<template-name>`
+* `cargo init <project-name> --template <crate-name>/<template-name>`
```

Comment:
The input to `cargo new` and `cargo init` are paths, not project names. Is this just a typo here?

Reply:
Oops! I was thinking project name in my head, not path. Great catch - I'll change that now

Reply:
Done!